src/state/block_tree.rs

```diff
@@ -0,0 +1,503 @@
+//! |Validator Set Update Block Height|Provided to [`initialize`](crate::replica::Replica::initialize).|
+//! |Validator Set Update Complete|Provided to [`initialize`](crate::replica::Replica::initialize).|
+//! |Locked QC|The [genesis QC](crate::hotstuff::types::QuorumCertificate::genesis_qc)|
+//! |Highest View Entered|0|
+//! |Highest Quorum Certificate|The [genesis QC](crate::hotstuff::types::QuorumCertificate::genesis_qc)|
+
+use crate::hotstuff::types::QuorumCertificate;
+use crate::pacemaker::types::TimeoutCertificate;
+use crate::types::basic::{ChildrenList, CryptoHash, ViewNumber};
+use crate::types::*;
+use std::cmp::max;
+use std::iter::successors;
+
+use self::basic::{AppStateUpdates, BlockHeight, Data, DataLen, Datum};
+use self::block::Block;
+use self::validators::{
+    ValidatorSet, ValidatorSetState, ValidatorSetUpdates, ValidatorSetUpdatesStatus,
+};
+
+use super::app_block_tree_view::AppBlockTreeView;
+use super::block_tree_snapshot::BlockTreeSnapshot;
+use super::kv_store::{KVGetError, KVStore};
+use super::write_batch::{BlockTreeWriteBatch, KVSetError};
+
+/// A read and write handle into the block tree exclusively owned by the algorithm thread.
+pub struct BlockTree<K: KVStore>(pub(super) K);
+
+impl<K: KVStore> BlockTree<K> {
+    pub(crate) fn new(kv_store: K) -> Self {
+        BlockTree(kv_store)
+    }
+
+    pub unsafe fn new_unsafe(kv_store: K) -> Self {
+        Self::new(kv_store)
+    }
+
+    /* ↓↓↓ Initialize ↓↓↓ */
+
+    pub fn initialize(
+        &mut self,
+        initial_app_state: &AppStateUpdates,
+        initial_validator_set_state: &ValidatorSetState,
+    ) -> Result<(), BlockTreeError> {
+        let mut wb = BlockTreeWriteBatch::new();
+
+        wb.apply_app_state_updates(initial_app_state);
+
+        let committed_validator_set = initial_validator_set_state.committed_validator_set();
+        let previous_validator_set = initial_validator_set_state.previous_validator_set();
+        let update_height = initial_validator_set_state.update_height();
+        let update_completed = initial_validator_set_state.update_completed();
+
+        wb.set_committed_validator_set(&committed_validator_set)?;
+        wb.set_previous_validator_set(&previous_validator_set)?;
+        if let Some(height) = *update_height {
+            wb.set_validator_set_update_block_height(height)?
+        }
+        wb.set_validator_set_update_completed(update_completed)?;
+
+        wb.set_locked_qc(&QuorumCertificate::genesis_qc())?;
+
+        wb.set_highest_view_entered(ViewNumber::init())?;
+
+        wb.set_highest_qc(&QuorumCertificate::genesis_qc())?;
+
+        self.write(wb);
+
+        Ok(())
+    }
+
+    /// Insert a block into the block tree. This includes:
+    /// 1. Setting pending app states and validator set updates associated with the block,
+    /// 2. Updating all relevant mappings from the block.
+    ///
+    /// # Precondition
+    /// [super::safety::safe_block]
```

Comment:
Is it okay to call `insert_block` twice with the same block? This should be documented in the preconditions section.

Line 182 (`siblings.push`) suggest that it is not ok.

Reply:
Will fix with https://github.com/parallelchain-io/hotstuff_rs/issues/44